File: make_site.rb
require "fileutils" # For copying files etc

# ["20210617", "20210618", "20210511", "20210428", "20210427", "20210609"]
date_strs = Dir.entries("forecasts").grep(/\d{8}/)

nice_dates_strs = []

date_strs.each do |date_str| # "20210617"
  year_str  = date_str[0...4] # "2021"
  month_str = date_str[4...6] # "06"
  day_str   = date_str[6...8] # "17"

  # Convert to integers
  year  = year_str.to_i  # 2021
  month = month_str.to_i # 6
  day   = day_str.to_i   # 17

  nice_date_str = "#{year}-#{month}-#{day}" # "2021-6-17"
  nice_dates_strs += [nice_date_str]

  forecast_date_dir = File.join("forecasts", date_str) # "forecasts/20210617"
  out_dir           = File.join("site", nice_date_str) # "site/2021-6-17"

  # Make directories if they don't exist yet.
  FileUtils.mkdir_p(out_dir)

  # e.g. ["t6z", "t0z", "t12z", "t18z"]
  run_hour_strs = Dir.entries(forecast_date_dir).grep(/t\d+z/)

  # [0, 6, 12, 18]
  run_hours = run_hour_strs.map { |str| str[/\d+/].to_i }.sort

  # Use a templating library. For this demo I'll just use string interpolation.

  out = ""
  out += "<html>\n"
  out += "<head><title>Nadocast #{nice_date_str}</title></head>\n"
  out += "<body>\n"
  out += "<h1>Nadocast #{nice_date_str}</h1>\n"

  run_hours.each do |run_hour|
    run_hour_dir = File.join(forecast_date_dir, "t#{run_hour}z") # "forecasts/20210617/t0z"

    out += "<h2>#{run_hour}Z Forecasts</h2>\n"

    # ["nadocast_conus_tor_20210617_t00z_f02-11.png", "nadocast_conus_tor_20210617_t00z_f02.png", "nadocast_conus_tor_20210617_t00z_f03.png", ...]
    png_names    = Dir.entries(run_hour_dir).grep(/\.png\z/i).sort
    png_names.each do |png_name|
      png_path     = File.join(run_hour_dir, png_name) # "forecasts/20210617/t0z/nadocast_conus_tor_20210617_t00z_f02-11.png"
      png_out_path = File.join(out_dir, png_name)      # "site/2021-6-17/nadocast_conus_tor_20210617_t00z_f02-11.png"

      FileUtils.cp(png_path, png_out_path) # This is slow for some unknown reason.
      out += "<img src=\"#{png_name}\">\n"
    end
  end

  out += "</body>\n"
  out += "</html>\n"

  puts out

  File.write(File.join(out_dir, "index.html"), out)
end

# Generate landing page.

out = ""
out += "<html>\n"
out += "<head><title>Nadocast</title></head>\n"
out += "<body>\n"
out += "<h1>Nadocast</h1>\n"

out += "<p>Eventually the latest forecast will go here.</p>\n"

nice_dates_strs.each do |nice_date_str|
  out += "<p><a href=\"#{nice_date_str}/index.html\">#{nice_date_str}</a></p>\n"
end

out += "</body>\n"
out += "</html>\n"

puts out

File.write(File.join("site", "index.html"), out)

File: make_site.js
//self notes: still get rid of that redundant variable
//I'm using forEach two different ways, maybe unnecessarily
//still need to sort pngs maybe
//Start here - need to move callback functions of all node functions and move them
//Q: will the / in '../main.css' always work?
//Q: not sure if there's any point in making templates .handlebars if using toString anyway.
//Q: make sure images are only copied over if they don't exist already? what about old ones that need to be deleted?
//Q: still unsure about let vs const
//Idk how webkit works

const fs = require('fs');
const path = require('path');
const Handlebars = require("handlebars");
const files = fs.readdirSync('forecasts');
const homeTemplate = fs.readFileSync('home_template.handlebars').toString();
const subTemplate = fs.readFileSync('subpage_template.handlebars').toString();

const nice_date_strs = [];

function newDir(date_str) {
    //Read dates from file names
    let year_str  = date_str.substr(0,4) // "2021"
    let month_str = date_str.substr(4,2) // "06"
    let day_str   = date_str.substr(6,2) // "17"

    //Convert dates into integers
    let year = Number(year_str); // 2021
    let month = Number(month_str); // 6
    let day = Number(day_str); // 17

    //Add formatted date to nice_date_strs array
    let nice_date_str = `${year}-${month}-${day}`;
    nice_date_strs.push(nice_date_str);

    //Specify path of forecast folders and to-be-created copies to later read & write
    let forecast_date_dir   =  path.join('forecasts', date_str); // "forecasts/20210617"
    let out_dir             =  path.join('site', nice_date_str); // "site/2021-6-17"

    //Make directories if they don't exist yet
    if (!fs.existsSync(path.join(__dirname,out_dir))){
        fs.mkdirSync(path.join(__dirname,out_dir));
    }

    //Initialize run_hours_strs to carry subfolder names like t0z 
    let run_hour_strs = []; 

    //Node reads & stores the folders in each bigger folder 
    //(e.g. t0z in forecasts/20210617)
    let folders = fs.readdirSync(path.join(__dirname,forecast_date_dir));
    folders.forEach(folder => {
        run_hour_strs.push(folder);
    });

    //Sort run_hour_strs by the number nested within
    run_hour_strs = run_hour_strs.sort(function(a,b) {
        let aVal    = Number(a.match(/\d+/)[0]);
        let bVal    = Number(b.match(/\d+/)[0]);
        if (aVal < bVal) { return -1;}
        if (aVal > bVal) { return 1;}
        else { return 0 }
    });

    //Create object to hold each subsection of images for a single subpage
    const run_hours_obj = {};

    //Copy over images and display them in succession on subpage
    function newSubDir(run_hour) {
        let run_hour_dir = path.join('/',forecast_date_dir, run_hour);        
        let png_names = [];
        let images = fs.readdirSync(path.join(__dirname,run_hour_dir));
    
        images.forEach((image) => {
            //console.log(image);
            if (/^.*[0-9]+-[0-9]+\.png.*$/i.test(image)) {
                png_names.unshift(image);
                //console.log("unshift");
                //console.log(image);
            }
            else if (/^.*\.png+$/i.test(image)) {
                png_names.push(image);
                //console.log("push");
            } 
        });

        //Create an object for a subsection of images to add to run_hours_obj
        let zlabel = `${run_hour.match(/\d+/)}Z Forecasts`;
        const run_hour_obj = {
            title: zlabel,
            images: png_names,
            imagesNum: png_names.length -1
        }
        run_hours_obj[zlabel] = run_hour_obj;

        //Create the path names with the png_names array, then copy them over
        png_names.forEach((png_name) => {
            let png_path = path.join(__dirname,run_hour_dir,png_name);
            let png_out_path = path.join(__dirname,out_dir,png_name);
            
            //Copy the actual images over and adding the image link to the html
            fs.copyFileSync(png_path,png_out_path);
        });
    }
    run_hour_strs.forEach(newSubDir);

    //Write the file using the subpage template
    const subTempComp = Handlebars.compile(subTemplate);
    const subOut = subTempComp({
        date: nice_date_str,
        Z: run_hours_obj
    });
    
    
    fs.writeFileSync(path.join(__dirname,out_dir,"index.html"),subOut);

}
files.forEach(newDir);


//Generate landing page.
const templateComplete = Handlebars.compile(homeTemplate);
const homeOut = templateComplete({
    subpage: nice_date_strs
});
console.log("Writing files");
fs.writeFileSync(path.join(__dirname,"site","index.html"),homeOut);